### PR TITLE
feat(dynamic-sampling): Implement sliding window

### LINKED ISSUE #48131: Sliding Window Fidelity
https://www.notion.so/sentry/Sliding-Window-Discussion-30ba1bc6d15349fd8faeb3b7ce1910cf

We are currently calculating the target fidelity rate based on the purchased volume of transactions (or actual usage - whichever is higher).
This has a few downsides:
- if customers buy more than they need, we'll give them an unnecessarily low sample rate
- on unlimited trials, the actual usage is always higher than the purchased volume of transactions (because that's unlimited so treated as 0) and creates a teeth saw effect in indexed transactions as org is sending more data and going constantly through lower and lower fidelity targets

We want to make the target fidelity logic based on the usage in the sliding window.
The initial proposal is as follows: 
- every 10 minutes, we check the root count of a project in the last 24 hours (root count is important, we need to know how many transactions this project is causing in other projects)
- we multiply that number to get the full month (so in this case *30 or *31?)
- based on that monthly number, we look up the target fidelity in our existing volume-to-sample-rate tables
- use that as a base sample rate for the project
- we do this on the project level (not the organization level); this will allow us to get rid of the Prioritize Low Volume Project Bias
- we feature flag this and, for now, only enable it for sentry
- make the size of the window (24 hours) and the frequency of how often we recalculate (10 minutes) configurable (option?), we'll likely iterate on those
- if you have this feature flag enabled, we'll probably need to disable the org fidelity calibration bias

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR implements the sliding window rebalancing algorithm which tries on a per-project basis to change the sampling tier based on forecasted volume.

The new implementation uses one flag for enabling the new sliding window behavior: `organizations:ds-sliding-window`.

In addition, it provides an option `dynamic-sampling:sliding_window.size` to control the window size in hours.

_This PR contains experimental code, which would likely increase our cogs but we are aware of that and will fallback to another approach in case this one will not turn out to be sensible enough._

Closes https://github.com/getsentry/sentry/issues/48131

### CODE REVIEW

**Review by iker-barriocanal — COMMENTED** [review]
The diff is quite big (almost 690 LOC), would it be possible to split this into several PRs? It's really hard to do a meaningful review in such a big diff.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (RaduW) [inline src/sentry/dynamic_sampling/sliding_window_rebalancing.py:64]: I think we can safely remove this.
It was initially done because we were not sure the DB will be able to cope with this query.

Since we are doing basically the same query in Prioritise Projects and we do a lot more for transactions we know there are no DB projects so there is no need (IMO) to bother with this.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (RaduW) [inline src/sentry/dynamic_sampling/sliding_window_rebalancing.py:91]: No need for this, you can safely remove it.

This clause made sense for retrieving transactions when we wanted to limit the number of records (i.e. transactions) returned for each (project,org).

There are no transactions here so you will have only one aggregation record for each ( proj, org) tuple.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (RaduW) [inline src/sentry/dynamic_sampling/sliding_window_rebalancing.py:42]: This functionality is fully contained in `fetch_projects_with_total_volumes` no need to re-write the same query.

`fetch_projects_with_total_volumes` in addition gets `keep` and `drop` counts which it gets wrongly... but you don't need them here.

After my fix of the rebalance bias we shouldn't need `fetch_projects_with_total_volumes` to return `drop` and `keep` so one of the two functions should go (i.e. either keep yours and use it for the project rebalance bias or the other way).

(My fault for not removing `keep` and `drop` from `fetch_projects_with_total_volumes`).

File: src/sentry/dynamic_sampling/tasks.py
Comment (RaduW) [inline src/sentry/dynamic_sampling/tasks.py:385]: NIT: since you never search in it there is really no need for a dictionary,
A List[Tuple[id, sampleRate]] would do (of course if you prefer it, dictionary works as well).

File: src/sentry/dynamic_sampling/tasks.py
Comment (RaduW) [inline src/sentry/dynamic_sampling/tasks.py:352]: I'm not very sure why we are doing this ( I stopped doing it with the org rebalancing task). 
We are creating a lot of Tasks on the Celery queue and I'm not so sure what's the benefit, since
the child task only gets the projects from cache and writes a hash in Redis.

Maybe check with other engineers on whether it makes sense .

File: src/sentry/options/defaults.py
Comment (RaduW) [inline src/sentry/options/defaults.py:692]: As I said before, not sure we need this ( I think we don't).
The query you are doing is already tested for performance and it has no issues.

**Review by RaduW — COMMENTED** [review]
Very good, and super fast.
I did not approve it because we need to sort out the query (it is effectively a duplicate of fetch_projects_with_total_volumes).

Other than the query everything looks good.
If it can wait until next week I'll cleanup fetch_projects_with_total_volumes ( or throw it away) and we can use only one of them. 
If we need this PR before please remove the `set_limitby` clause  ( and probably get rid of sample rate modulo... ) and we'll clean up the duplication latter

File: src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py:38]: This is a very good point but since billing is monthly (to my understanding), it's better if we align to that.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/sliding_window_rebalancing.py:64]: Will do. I have put it in place since maybe we wanted a gradual rollout but since this query is even less expensive than the one before, I don't expect any regressions.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/sliding_window_rebalancing.py:42]: I would prefer to refactor in a followup PR, since I don't want to touch any code related to prioritise projects. I will add a TODO and a ticket.

File: src/sentry/dynamic_sampling/tasks.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/tasks.py:352]: I worked under the assumption that this parallelization of work was designed after careful consideration. Considering the complexity of the job I would also keep it in the same, especially since we will not need the `Project.objects.get_many_from_cache` once I merge the updated `get_blended_sample_rate`.

File: src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py:29]: "forecasted" is probably a misleading term here, this is merely extrapolated.

File: src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py
Comment (jan-auer) [reply src/sentry/dynamic_sampling/rules/helpers/sliding_window_rebalancing.py:38]: It would be interesting to make the window configurable internally (as sentry option) and play with different widths ranging from an hour to 3 days. This allows us to balance towards responsiveness or robustness depending on what we see with real world data.

It's fine to keep this initial implementation hard-coded to one day, however. If we choose to test lower than one day, we'd also have to move to a lower granularity in the query.

File: src/sentry/dynamic_sampling/sliding_window_rebalancing.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/sliding_window_rebalancing.py:104]: Can we add an error message here?

File: src/sentry/dynamic_sampling/rules/base.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/rules/base.py:13]: _nit_ on naming: This is no rebalancing, this computes the base sample rate

File: src/sentry/dynamic_sampling/tasks.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/tasks.py:126]: `get_blended_sample_rate` can take either an org or a project, see the doc string on that method. Please verify with the implementation in getsentry, but based on the signature description, you can run:

```py
desired_sample_rate = quotas.get_blended_sample_rate(project=None, organization_id=org_volume.org_id)
```

File: src/sentry/dynamic_sampling/tasks.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/tasks.py:157]: Please interpolate the min and max values into the error message

File: src/sentry/dynamic_sampling/tasks.py
Comment (jan-auer) [reply src/sentry/dynamic_sampling/tasks.py:352]: Adding on to Radu's comment, it could make sense to deliberately run everything in one task, so that we can also control the order:
- first, update the base rates through sliding window
- then, run all other factor queries
- finally, run the correction task, already factoring in the updated base rates

fine to keep as is for the first go, but please consider this during cleanup

File: src/sentry/dynamic_sampling/tasks.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/tasks.py:384]: If we have no volume, should we instead bump the sample rate to 100%? That would help with onboarding, where we usually want to see the first event through.

If `None` here means something else, like we were unable to retrieve stats etc, then we shouldn't do this, of course.

File: src/sentry/dynamic_sampling/tasks.py
Comment (jan-auer) [inline src/sentry/dynamic_sampling/tasks.py:396]: What are reasons this returns `None` in practice? If it's just the mentioned case that DS is enabled for this org/project, then we can keep this. But if there are other cases, we should make sure that internally an error is logged to give us a way to debug this. That error log should go in getsentry.

File: src/sentry/dynamic_sampling/tasks.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/tasks.py:126]: I have updated the signature of `get_blended_sample_rate` yesterday. I did not update it in the previous usages across Sentry. I will do it in a separate PR. The code you are seeing here has only been moved.

File: src/sentry/dynamic_sampling/tasks.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/tasks.py:396]: We actually already had some of these cases for the `get_blended_sample_rate` and we assumed it was a consistency problem on the billing side.

Comment (iambriccardo) [reply src/sentry/dynamic_sampling/tasks.py:396]: We will investigate further in case error samples increase.

File: src/sentry/dynamic_sampling/tasks.py
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/tasks.py:384]: This function will return `None` in case of errors. The value will be always > 0, assuming a positive root count and a value > 0 which for our query is.

**Review by jan-auer — APPROVED** [review]

**Pull request merged**
